ORBIT/core/environment.py

```diff
@@ -0,0 +1,238 @@
+        """
+
+        ws = {}
+        for k in list(constraints.keys()):
+            if "windspeed" in k:
+                ws[k] = constraints.pop(k)
+
+        if not ws:
+            return constraints
+
+        if "windspeed" in self.state.dtype.names:
+            if len(ws) > 1:
+                raise ValueError(
+                    "Multiple constraints applied to the 'windspeed' column."
+                )
+
+            return {**constraints, "windspeed": list(ws.values())[0]}
+
+        for k, v in ws.items():
+
+            if k == "windspeed":
+                height = self.simplify_num(self.default_height)
+
+            else:
+                val = k.split("_")[1].replace("m", "")
+                height = self.simplify_num(val)
+
+            name = f"windspeed_{height}m"
+            if name in self.state.dtype.names:
+                pass
+
+            else:
+                loc = bisect(self.ws_heights, height)
+                if loc == 0:
+                    self.extrapolate_ws(self.ws_heights[0], height)
+
+                elif loc == len(self.ws_heights):
+                    self.extrapolate_ws(self.ws_heights[-1], height)
+
+                else:
+                    h1 = self.ws_heights[loc - 1]
+                    h2 = self.ws_heights[loc]
+                    self.interpolate_ws(h1, h2, height)
+
+            constraints[name] = v
+
+        return constraints
+
+    @property
+    def ws_heights(self):
+        """Returns heights of available windspeed profiles."""
+
+        columns = [c for c in self.state.dtype.names if "windspeed" in c]
+
+        heights = []
+        for c in columns:
+            val = c.split("_")[1].replace("m", "")
+            heights.append(float(val) if "." in val else int(val))
+
+        return sorted(heights)
+
+    def interpolate_ws(self, h1, h2, h):
+        """
+        Interpolates between two windspeed profiles using the power law. This
+        method will calculate the power law coefficient between `h1` and `h2`.
+
+        Parameters
+        ----------
+        h1 : int | float
+            Lower measurement height.
+        h2 : int | float
+            Upper measurement height.
+        h : int | float
+            Desired profile height.
+        """
+
+        ts1 = self.state[f"windspeed_{h1}m"]
+        ts2 = self.state[f"windspeed_{h2}m"]
+        alpha = np.log(ts2.mean() / ts1.mean()) / np.log(h2 / h1)
+        print(alpha)
+        ts = ts1 * (h / h1) ** alpha
+
+        self.state = np.array(append_fields(self.state, f"windspeed_{h}m", ts))
+
+    def extrapolate_ws(self, h1, h):
+        """
+        Extrapolates a windspeed profile using the power law coefficient
+        at `self.alpha`.
+
+        Parameters
+        ----------
+        h1 : int | float
+            Measurement height.
+        h : int | float
+            Desired profile height.
+        """
+
+        ts1 = self.state[f"windspeed_{h1}m"]
+        if h > h1:
```

Comment:
Power law should work whether h > h1 or h < h1 - why separate equations?  Can this be simplified?

Reply:
I was going off some old code I had that had two equations but good point. I'll modify this as well.